### PR TITLE
kernel: msgq: Add k_msgq_peek() API

### LINKED ISSUE #10948: k_msgq API to peek the message at the head of the queue
Currently there are 2 primary APIs for the kernel message queues. _put() and _get().
In a few drivers, I could have avoided creating additional message queues if there was an API to read the message at the head of the queue without actually deleting the message from the queue.
For example while tracking data buffers through the driver from 
- application to the driver
- within driver to the DMA descriptor
- within driver from DMA descriptor to ISR
- within driver from ISR to application callback

Since the message gets deleted from the queue upon _get(), another queue needs to be created for the different "hand-offs" within the driver.

What do people think of having a _peek() API for message queues?

### PR DESCRIPTION
**Pull request by sathishkuttan**
Adds API, implementation and updates documentation for a k_msgq_peek API

Resolves #10948

### CODE REVIEW

File: doc/kernel/data_passing/message_queues.rst [hunk 1]
```diff
@@ -147,6 +153,29 @@ to process data items generated by one or more producing threads.
         }
     }
 
+Peeking into a Message Queue
+============================
+
+A data item is read from a message queue by calling :cpp:func:`k_msgq_peek()`.
+
+The following code peeeks into the message queue to read the data item at the
```
Comment (andyross) [inline doc/kernel/data_passing/message_queues.rst]: spellling

**Comment by nashif** [comment]
@sathishkuttan we need a test for this as well. thanks for looking into that.

**Review by andyross — APPROVED** [review]
Looks correct.   I'm a little confused about the utility here, though.  This is a copying data structure to begin with.  Doing I/O on it with a peek means taking the overhead of an **extra** copy to eventually remove the item, which.... really sounds to me like you want to be using a different data structure in the first place.

**Comment by sathishkuttan** [comment]
> I'm a little confused about the utility here, though

Issue #10948 describes the need/utility.

File: doc/kernel/data_passing/message_queues.rst (hunk 1 above)
Comment (sathishkuttan) [reply doc/kernel/data_passing/message_queues.rst]: fixed spelling.

**Comment by sathishkuttan** [comment]
> we need a test for this as well

what would be the appropriate place to add a test?
Is there an existing file with tests for `k_msgq_put()` and `k_msgq_get()`?

File: doc/kernel/data_passing/message_queues.rst [hunk 2]
```diff
@@ -55,6 +55,12 @@ Any number of receiving threads may wait simultaneously when the ring buffer
 is empty; when a data item becomes available it is given to
 the highest priority receiving thread that has waited the longest.
 
+A data item can also be **read** from a message queue without removing
```
Comment (dbkinder) [inline doc/kernel/data_passing/message_queues.rst]: I'd simplify and replace this sentence (three lines) with:

    You can also peek at the message on the head of a message queue
    without removing it from the queue.

**Review by dbkinder — DISMISSED** [review]

**Comment by nashif** [comment]
> Is there an existing file with tests for `k_msgq_put()` and `k_msgq_get()`?

tests/kernel/msgq/msgq_api

**Review by tyrantlizard77 — DISMISSED** [review]
please add some test code that proves this API works

File: include/kernel.h [hunk 3]
```diff
@@ -3327,6 +3327,23 @@ __syscall int k_msgq_put(struct k_msgq *q, void *data, s32_t timeout);
  */
 __syscall int k_msgq_get(struct k_msgq *q, void *data, s32_t timeout);
 
+/**
+ * @brief Peek/read a message from a message queue.
+ *
+ * This routine reads a message from message queue @a q in a "first in,
+ * first out" manner and leaves the message in the queue.
+ *
+ * @note Can be called by ISRs, but @a timeout must be set to K_NO_WAIT.
```
[the remark this answers is not in the document]
Comment (sathishkuttan) [reply include/kernel.h]: > there is no timeout, delete this line

deleted.

**Review by nashif — APPROVED** [review]

**Pull request merged**